services/src/util/config.rs

```diff
@@ -0,0 +1,113 @@
+use std::sync::RwLock;
+
+use crate::error;
+use crate::error::Result;
+use config::{Config, File};
+use lazy_static::lazy_static;
+use serde::Deserialize;
+use snafu::ResultExt;
+use std::path::Path;
+
+lazy_static! {
+    static ref SETTINGS: RwLock<Config> = RwLock::new({
+        let mut settings = Config::default();
+        // test may run in subdirectory
+        #[cfg(test)]
+        let paths = [
+            "../Settings-default.toml",
+            "Settings-default.toml",
+            "../Settings-test.toml",
+            "Settings-test.toml",
+        ];
+
+        #[cfg(not(test))]
+        let paths = ["Settings-default.toml", "Settings.toml"];
+
+        #[allow(clippy::filter_map)]
+        let files: Vec<File<_>> = paths
+            .iter()
+            .map(Path::new)
+            .filter(|p| p.exists())
+            .map(File::from)
+            .collect();
+
+        settings.merge(files).unwrap();
+
+        settings
+    });
+}
+
+pub fn get_config<'a, T>(key: &str) -> Result<T>
```

Comment:
will this de-serialize the config on each call?

Reply:
good point. Yes it deserializes on each call. Fortunately, it doesn't read the file each time ;)

It also only deserializes lazily once the config key is accessed. So we might end up with the problem, where a faulty configuration is not detected until the value is actually used for the first time.

Maybe we should just bundle all the configs in a Config struct and deserialize into it from the config file once at startup. Then we at least know that all required values are present and valid. I am not sure if we even need the config crate in this case. It makes it at least easy to merge multiple config files like Settings-default.toml and Settings.toml

Reply:
We could deserialize into a typed map/list. Then we can keep the get<T> thing. I also think that a validation of the config when it is loaded is important.

Reply:
but what ist the type of the value of the map? `Map<String, ?>`. Do we put all config structs in an enum?

Reply:
we will have to use a crate like https://docs.rs/typemap/0.3.3/typemap/ or build something similar

Reply:
I created a task for this https://dbs-projects.mathematik.uni-marburg.de/T1227